### PR TITLE
AEA-516 Move functionalities from cli generate to generator

### PR DESCRIPTION
**Pull request by 5A11**
This PR includes the following:

- Moved the following functionalities from CLI generate to the generator:
  - Checking `protoc` is installed
  - Checking `black` is installed
  - Loading protocol specification yaml file
  - Run black code formatting
- Added the following functionalities:
  - Checking whether a protobuf schema file is valid, by trying to compile it.
  - The above process is now atomic, if invalid it rolls back everything generated. If any errors, it correctly forwards them as part of exception message.
- Running `protoc` is now via `subprocess` instead of `os.system(cmd)`
- `extract_specification.py`, `validate.py`, `common.py` pass `pylint` now
- Updating tests to reflect the above changes

## Types of changes

What types of changes does your code introduce to agents-aea?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have checked that code coverage does not decrease.
- [ ] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/cli/generate.py [hunk 1]
```diff
@@ -20,9 +20,6 @@
 """Implementation of the 'aea generate' subcommand."""
```
Comment (5A11) [inline aea/cli/generate.py:20]: Code related to `Black` and `protoc` are now in the generator as they are needed by the generator and not just by the generate cli command.

File: aea/protocols/generator/base.py [hunk 2]
```diff
@@ -183,24 +189,50 @@ class ProtocolGenerator:
...
 
-        :param protocol_specification: the protocol specification object
+        :param path_to_protocol_specification: path to protocol specification file
         :param output_path: the path to the location in which the protocol module is to be generated.
         :param path_to_protocol_package: the path to the protocol package
 
         :return: None
         """
-        self.protocol_specification = protocol_specification
+        self.path_to_protocol_specification = path_to_protocol_specification
```
Comment (DavidMinarsch) [inline aea/protocols/generator/base.py:205]: Now this method does even more; please isolate functionalities...e.g. have individual methods which wrap separate operations and then call them from here. Some of these are not related to the object so they can be helper methods

File: aea/protocols/generator/common.py [hunk 3]
```diff
@@ -108,3 +116,90 @@ def _get_sub_types_of_compositional_types(compositional_type: str) -> tuple:
...
+        return True, "protobuf file is valid"
+    except subprocess.CalledProcessError as e:
+        pattern = name + ".proto:[0-9]+:[0-9]+: "
+        error_message = re.sub(pattern, "", e.stderr[:-1])
+        return False, error_message
+    except Exception:
+        raise
+
+
+def run_protoc(path_to_generated_protocol_package, name) -> subprocess.CompletedProcess:
```
Comment (DavidMinarsch) [inline aea/protocols/generator/common.py:180]: should call this `try_run_protoc`

File: aea/protocols/generator/common.py [hunk 4]
```diff
@@ -108,3 +116,90 @@ def _get_sub_types_of_compositional_types(compositional_type: str) -> tuple:
...
+    :return: A ProtocolSpecification object
+    """
+    config_loader = ConfigLoader(
+        "protocol-specification_schema.json", ProtocolSpecification
+    )
+    protocol_spec = config_loader.load_protocol_specification(open(specification_path))
+    return protocol_spec
+
+
+def run_black_formatting(path_to_protocol_package: str) -> None:
```
Comment (DavidMinarsch) [inline aea/protocols/generator/common.py:135]: `try_run_black_formatting`

File: tests/test_cli/test_generate/test_generate.py [hunk 5]
```diff
@@ -57,20 +57,21 @@ def test__generate_item_file_exists(self, *mocks):
...
+    @mock.patch("aea.protocols.generator.base.shutil.which", _which_mock)
     def test__generate_item_no_res(self, *mocks):
         """Test for fetch_agent_locally method no black."""
         ctx_mock = ContextMock()
         with self.assertRaises(ClickException) as cm:
             _generate_item(ctx_mock, "protocol", "path")
         expected_msg = (
-            "Please install black code formater first! See the following link: "
+            "There was an error while generating the protocol. The protocol is NOT generated. Exception:\n"
+            "Cannot find black code formatter. To install, please follow this link: "
```
Comment (DavidMinarsch) [inline tests/test_cli/test_generate/test_generate.py:68]: Long error messages should be avoided.

File: aea/protocols/generator/common.py [hunk 6]
```diff
@@ -108,3 +116,90 @@ def _get_sub_types_of_compositional_types(compositional_type: str) -> tuple:
...
+                "{}/{}.proto".format(path_to_generated_protocol_package, name),
+            ],
+            capture_output=True,
+            text=True,
+            check=True,
+            env=os.environ.copy(),
+        )
+    except Exception:
+        raise
+    return protoc_process
```
[the remark this answers is not in the document]
Comment (5A11) [reply aea/protocols/generator/common.py:205]: Done

File: aea/protocols/generator/common.py (hunk 3 above)
Comment (5A11) [reply aea/protocols/generator/common.py:180]: Then we should change all similar functions that run a programme (almost always within a try), `try_to_run_...`. Like `run_black` -> `try_to_run_black`.

File: aea/protocols/generator/common.py (hunk 4 above)
Comment (5A11) [reply aea/protocols/generator/common.py:135]: Yep

File: aea/protocols/generator/common.py (hunk 3 above)
Comment (5A11) [reply aea/protocols/generator/common.py:180]: Done

File: aea/protocols/generator/common.py (hunk 4 above)
Comment (5A11) [reply aea/protocols/generator/common.py:135]: Done

File: tests/test_cli/test_generate/test_generate.py (hunk 5 above)
Comment (5A11) [reply tests/test_cli/test_generate/test_generate.py:68]: Done

File: aea/protocols/generator/base.py (hunk 2 above)
Comment (5A11) [reply aea/protocols/generator/base.py:205]: Done

File: aea/protocols/generator/common.py [hunk 7]
```diff
@@ -108,3 +116,90 @@ def _get_sub_types_of_compositional_types(compositional_type: str) -> tuple:
...
+    try:
+        run_protoc(path_to_generated_protocol_package, name)
+        os.remove(os.path.join(path_to_generated_protocol_package, name + "_pb2.py"))
+        return True, "protobuf file is valid"
+    except subprocess.CalledProcessError as e:
+        pattern = name + ".proto:[0-9]+:[0-9]+: "
+        error_message = re.sub(pattern, "", e.stderr[:-1])
+        return False, error_message
+    except Exception:
+        raise
```
[the remark this answers is not in the document]
Comment (5A11) [reply aea/protocols/generator/common.py:177]: I always reraise the exception to higher level, to get caught by the cli generate command to present the appropriate message. 

In this case, Pylint complained for `subprocess.run`, `check` must be `True`. 
This makes it so if `protoc` returns any syntax error on the protobuf schema file, `subprocess` creates a special `CalledProcessError` exception containing attributes such as `stderr`. So originally the code was something like this:

```
    try:
        process = try_run_protoc(path_to_generated_protocol_package, name)
        if process.stderr == "":
            os.remove(os.path.join(path_to_generated_protocol_package, name + "_pb2.py"))
            return True, "protobuf file is valid"
        else:
            pattern = name + ".proto:[0-9]+:[0-9]+: "
            error_message = re.sub(pattern, "", e.stderr[:-1])
            return False, error_message
    except Exception:
        raise
```
To make Pylint pass, the `else` code was moved to `CalledProcessError` exception handling.
